pkg/storegateway/sharding_strategy.go

```diff
@@ -159,18 +180,33 @@ func GetShuffleShardingSubring(ring *ring.Ring, userID string, limits ShardingLi
 type shardingMetadataFilterAdapter struct {
 	userID   string
 	strategy ShardingStrategy
+
+	// Keep track of the last blocks returned by the Filter() function.
+	lastBlocks map[ulid.ULID]struct{}
 }
 
 func NewShardingMetadataFilterAdapter(userID string, strategy ShardingStrategy) block.MetadataFilter {
 	return &shardingMetadataFilterAdapter{
-		userID:   userID,
-		strategy: strategy,
+		userID:     userID,
+		strategy:   strategy,
+		lastBlocks: map[ulid.ULID]struct{}{},
 	}
 }
 
 // Filter implements block.MetadataFilter.
+// This function is NOT safe for use by multiple goroutines concurrently.
```

Comment:
It's not used concurrently. Not even Thanos filters are thread-safe. I've added the comment just to make it clear.